### ISSUE TITLE
[Domino Visual Editor] View Column Properties

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1017]
We would like to add support for Shared columns in the Domino View editor (#1016), so that we can reuse columns between forms.

**Comment by JoelProminic** [comment]
Here are some screenshots from Domino Designer.  I highlighted the properties that I use the most - we should focus on these initially:
![image](https://user-images.githubusercontent.com/24532731/235005577-c70412c2-9ba1-4852-8d7f-825b3da843f6.png)
![image](https://user-images.githubusercontent.com/24532731/235005767-cd9f0ee2-d908-4143-bc60-5c14be2f835f.png)
![image](https://user-images.githubusercontent.com/24532731/235005882-af6aa7af-6f29-41d7-b1a0-13ed2146f49a.png)![image](https://user-images.githubusercontent.com/24532731/235005916-99ce3cb9-9945-44e3-82ca-1ba25d498147.png)![image](https://user-images.githubusercontent.com/24532731/235006001-1d98b46f-8e8e-4113-bece-e788b2b34fd5.png)![image](https://user-images.githubusercontent.com/24532731/235006057-db75bb72-0629-4f83-a8bc-6b61888cca92.png)
![image](https://user-images.githubusercontent.com/24532731/235006373-594410b3-8b2c-45cf-b7ce-dfbc3fa13395.png)


Font:
Title:
![image](https://user-images.githubusercontent.com/24532731/235006160-ad5cd7dd-a469-42c4-8ad4-2ddac237ecb6.png)
Values:
![image](https://user-images.githubusercontent.com/24532731/235005822-1649810c-6cbf-4365-8247-26e3475746c8.png)

**Comment by JoelProminic** [comment]
The options for "Use preferences from" in the Date Format section are:
- User setting
- Custom

"Custom" enables additional fields:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/43a32c80-dec0-4576-89e0-f595074b384c)

**Comment by JoelProminic** [comment]
Some more formatting options

### Date format

#### Display Date

Show
- Only month, day and year
- Only weekday, month and day
- Only month and year
- Only month and day
- Only year

Special (multi-select or checkbox field):
- Show 'today' when appropriate
- Always show 4 digit year
- Show 4 digit year for 21st century
- Show year only if not this year

Calendar:
- Gregorian
- Hijri

Format:
- YMDW
- WMDY
- WDMY

Day:
- d
- dd

Month:
- m
- mm
- mmm
- mmmm

Year:
- yy
- yyyy

Weekday:
- w
- ww
- (w)
- (ww)

#### Display Time

Show
- All
- Hours, minutes, and seconds
- Hours and minutes
- Hours only

Time zone
- Adjust time to local zone
- Always show time zone
- Show only if zone not local

Format
- 12 hour
- 24 hour

**Comment by feather812002** [comment]
Hello, Joel
I got some trouble about the different between the numberformat value list on the HCL document and the Designer UI,
As you see, the number format form the Designer UI ,it include  :   Decimal, Percent, Scientific,Currency,Bytes(K/M/S)
But from the HCL online document the number format only contain :  `<!ENTITY % numberformat.formats "general | fixed | scientific | currency">`
I confirm if I input other value ,it will compile error like you report on last day,
So, I think we need more details about what format when user setting the number as "Bytes", Now, in the Moonshine-IDE, when user select the "Decimal", I setting it as "fixed", it can success compile , and open the view success in the Notes client, but I got some stranger display value , the value is "12345.67" , but it always show as "1.00" in the view , I don't why it will show this ?
I already try "12345.67" and "@TextToNumber("12345.67")"
Nothing change , it just "1.00"

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/9aa9a79f-5323-4de2-abdc-15b709dab361)

**Comment by JoelProminic** [comment]
I created a test form to check  the numberformat properties:

[2023_06_16__ColumnFormats.dxl.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/11771709/2023_06_16__ColumnFormats.dxl.zip)

Label | `format` | Other Properties
-- | -- | --
Decimal | fixed | 
Percent | fixed |  percent=true
Scientific | scientific | 
Currency | currency | 
Bytes | fixed |  bytes=true

**Comment by JoelProminic** [comment]
I did some debugging with the numberformat value issue, and I found that I could fix the problem by moving the `numberformat` element to be between `columnheader` and `code`:

```
  <column sort="none" categorized="false" hidedetailrows="false" width="9" resizable="true" sortnoaccent="true" sortnocase="true" showaslinks="false" itemname="$1">
    <!-- original position - value=1 and header is blank -->
    <font style="normal" color="#000000"/>
    <columnheader title="TestDecimal">
      <font name="sans-serif" style="normal" size="9" color="#000000"/>
    </columnheader>
    <numberformat format="fixed" digits="1" punctuated="false" parens="false" percent="false" bytes="false" decimalsym="." thousandssep="," currencysymtype="custom" currencysym="$" usecustomsym="false"/>
    <code event="value">
      <formula>12345.67</formula>
    </code>
  </column>
```

**Comment by JoelProminic** [comment]
Currency options.  Note that these are only available if "Use preferences from" is set to "Custom":

The lists of currency symbols and country abbreviations are very long, so we need to look for a list somewhere:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/33d28918-f508-468d-8a43-9a728236f5e6) ![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/034ea308-513c-4b3d-971e-49d064b2ba66)

**Comment by JoelProminic** [comment]
Here are some examples of the DateTime format.  I tried to make the selected options clear from the column name, but these options are more complex.  Let me know if you need clarification:
[2023_06_22__TestColumnFormat_DateTime.dxl.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/11835694/2023_06_22__TestColumnFormat_DateTime.dxl.zip)

The custom columns were especially complicated, so here are some screenshots:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/88d8e851-d158-424d-ba0c-d2741529c79a) ![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/4bbc5a13-f04d-4fa8-a6f2-863b5afe8f4a)

**Comment by JoelProminic** [comment]
I populated more values for Month, Year, and Weekday here:  https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1017#issuecomment-1591462549

I did another test with Display Abbreviated Date, but I still don't see a clear property for this in the DXL.  I suspect this sets a collection of properties instead, like a set of common properties.

**Comment by JoelProminic** [comment]
I noticed that the Names format options were not preserved when reopening a view:
1. Create a new view
2. Change the format for the UNID column to "Names" and change all the properties
3. Close and reopen the form
4. Check the column format.  BUG: all properties have reset to defaults

Otherwise, I saw that the Names format properties were set like I expected in the DXL, and the database compiled without any errors.

**Comment by JoelProminic** [comment]
(NOTE:  I tried to add this column earlier, but it was apparently not submitted properly)
Similarly, I see that these DateTime properties are not preserved if I reopen the form.  I am testing with build 809.
- Date Separators - I do see that they remain in the Sample date, though.
- Time Separators


I did some tests with the Sample date, here are some problems
- The time is still displayed after I disable Display Time
- Disabling Display Date changes the date format to `2023 07/07/ 16:10:02` when the date should not be displayed (also note the misplaced `/`s).  This format does not revert when I re-enable Display Date
- Display Time > Show options leave trailing ":".  For example "Hours and Minutes" > `2023 07/07/ 16:37:`

Date > Show | Current | Domino Designer, default user preferences
-- | -- | --
"All" | missing option | `Fri 07/07/2023`
"Only month, day and year" | `2023 07/07/` | `07/07/2023`
"Only weekday, month, and day" | `07/07/Fri` | `Fri 07/07`
"Only month and year" | `2023 07/` | `07/2023`
"Only Month and Day" | `2023 07/07/Fri ` | `07/07`
"Only Year" | `2023 ` (extra whitespace) | `2023`

**Comment by JoelProminic** [comment]
I noticed some remaining problems with the font setting today.
1. Create a new view and select the UNID column
2. Check the Title and Column font.  BUG:  both show "monospace".  The default should be "sans-serif" instead.
3. Change the Title Font to "serif"
4. Check the Column font.  BUG:  this also changed to "serif"
5. Save
6. Reopen the form
7. Check the font.  BUG:  both the Title and Column show "monospace" for the font.

**Comment by JoelProminic** [comment]
We still need to return to the shared columns for this issue, but the current interface would be useful for some agent and database development work I've been doing lately.  So, I'd like to get this branch merged before we start on shared actions (or work on them in a separate branch).   I'll consider whether I want to create a new issue - we ended up focusing on implementing the column properties in this issue.

**Comment by JoelProminic** [comment]
When I duplicate a form, I see that the name doesn't update in the .view XML.  Rename also has this problem.  For both cases, `<view name="-viewname`"` should be updated with the new view name.

Do we have intermediate XML files for views.  I didn't find these in visualeditor-src when I was checking the Duplicate and Rename actions.

**Comment by feather812002** [comment]
@JoelProminic We no intermediate XML files for views, it directly operate the node in the dxl .

**Comment by JoelProminic** [comment]
I moved the shared column logic to #1216.  This issue, which covers the basic view properties, can be merged along with #1016.
